### ISSUE TITLE
Support for shapes

### ISSUE THREAD

**Issue by mitar** [numpy/numpy_stubs#5]
See [how contracts package](https://github.com/AndreaCensi/contracts) are trying to provide support for something similar to shapes. They are extending annotations in a different way than just standard typing and maybe something like that could be also done. So instead of providing specific extension (PEP) for typing to allow thing like shapes, it might be maybe more useful to determine a syntax for general constraints on types and use that, in addition to standard types through typing.

**Comment by ethanhs** [comment]
This will take quite some work here and in mypy. At a discussion held several weeks ago, the agreement was on the plan in the [README](https://github.com/numpy/numpy_stubs/blob/master/README.md). So while this will happen, adding support for checking shapes will not happen soon. There is much to decide, such as how to define syntax related to declaring array shape, and how an actual type checking implementation should work.

**Comment by rmcgibbo** [comment]
@shoyer: I'm not sure where this should be noted, but the framing of #6 and python/typing#516 seems to indicate that you're favoring putting the full shape in the type specification: that is, for a two-dimensional array of size (M,N), the proposal seems to be that the type annotation would convey both the value of M and N.

I think we should consider a more modest version in which only the number of dimensions is tracked via the type, but not the shape itself. That is, instead of `ndarray[Shape[100, 200], Dtype]` for an array of shape (100,200), we go for `ndarray[2, Dtype]`, to indicate only the fact that the array has ndim=2.

**Comment by rmcgibbo** [comment]
Tracking only the rank (as opposed to the full shape) of the array through the type system is, for what it's worth, how all of the C++ tensor libraries work to the best of my knowledge. It's more consistent with the rest of the PEP484 container types as well, in the sense that the type `List[int]` doesn't track the number of elements.

**Comment by shoyer** [comment]
@rmcgibbo Yes, this needs justification.

My reasoning for favoring specifying shapes in types comes down to:
- Rank and shape are not mutually exclusive: if we do shape, we get rank as well. We could even have an explicit `Rank` type, which we tell type checkers should be treated as equivalent to anonymous dimensions, e.g., `ndarray[Rank[2]]` could be the same as `ndarray[Shape[:, :]]`.
- For code directly using NumPy, rank checks would not be anywhere near as valuable as shape checks. The problem is that nearly NumPy operation is valid on inputs with an arbitrary number of dimensions (even between arguments, with broadcasting). Giving up on shape checks would means we couldn't type ufuncs in a useful way, or catch the common errors with broadcasting. This would basically mean reserving shape/rank checks for the (many) applications and libraries that only use on particular ranks (e.g., vectors and matrices).

With regards to C++ libraries, [Eigen](https://eigen.tuxfamily.org/dox/group__TutorialMatrixClass.html) at least does allow for either runtime and statically defined shapes. That said, I certainly would not bother with static/compile-time shapes on their own, because it is indeed rare that you know the exact shape of arrays before running a program. But with generics, we have the possibility of doing much more sophisticated pattern matching. This isn't feasible in C++, where you would need to use templates for every possible dimension size. In contrast, we can probably do something more sophisticated in Python to allow for typing shapes without enumerating all of them (similar to what you see in Haskell).

**Comment by rmcgibbo** [comment]
Shape is clearly more expressive than rank. So in that sense, I agree with both of your bulleted points above (not mutually exclusive, and shape checks provide more value to users). There are two factors on the other side that come to mind.

* In almost any real code, it will be impossible for mypy to statically infer the shapes. For example, take the following:

```
M = 5
N = 10
arr = np.zeros((M, N))
reveal_type(arr)
```

Mypy tracks the type of M and N, but not their values. It can infer the rank of `arr`, but not the full shape. Want to do something more complicated?

```
def compute_length() -> int:
    return <something>
def compute_width() > int:
    return <something>
arr = np.zeros((compute_length(), compute_width())
reveal_type(arr)
```
Not going to work.

* If the shape is to be tracked explicitly, for example, to the best of my understanding there is no way to express even the *simplest* numpy function signatures with the current generation of PEP484/mypy. For example, take

```
# requires a type-level addition function to express the return type.
def concatenate_1d_arrays(arr1 : ndarray[D, Shape[M]], arr2 : ndarray[D, Shape[N]] -> ?
```

**Comment by shoyer** [comment]
Take another look at [my doc on shape typing](https://docs.google.com/document/d/1vpMse4c6DrWH5rq2tQSx3qwP_m_0lyn-Ij4WHqQqRHY/edit#heading=h.rkj7d39awayl). You're right that functions like `concatenate` are hard, but I would claim that even mere matching of integer sizes (without arithmetic) along with dimension identity would be enough to be valuable:
- You could catch broadcasting errors when using ufuncs.
- You could catch errors with regards to which dimension is removed by aggregation or indexing
- You could verify that the shapes of non-concatenated axes of arrays in `concatenate` match, and are propagated to the result, e.g., `stack_matrices(x: ndarray[Shape[:, N], y: ndarray[Shape[:, N]]) -> ndarray[Shape[:, N]]`.

> In almost any real code, it will be impossible for mypy to statically infer the shapes. For example, take the following:
```
M = 5
N = 10
arr = np.zeros((M, N))
reveal_type(arr)
```

I'm not certain that mypy will be unable to infer shapes in this case. If typing adds support for literals (which is almost certainly necessary regardless for NumPy), then I imagine it will also support literals in variables, in the same way that it supports type aliases.

**Comment by rmcgibbo** [comment]
My suspicions are that shape tracking is (1) going to add a small marginal value, since it will be rarely be possible to statically infer them (e.g. my two examples above) in real world code, whereas it will often be possible to statically infer rank in real world code; and (2) going to add a ton of complexity.

If there's a way to design the syntax so that it is initially rank-only, build a viable package around that, and add shape information afterwards in a second stage, then that would be my preference.

**Comment by hameerabbasi** [comment]
Maybe for v1.0 we could have `ndarray[2, Dtype]` be an alias for `ndarray[Shape[Any, Any], Dtype]`, ignoring the length of the shape but keeping the `ndim` in mind. This would probably be a good-ish idea to keep forward compatibility.

**Comment by shoyer** [comment]
Yes, I agree that it could make sense to add shape support incrementally. We'll probably still need  upstream work in typing even to make that work well, though (e.g., to support integer values in types).
